### ISSUE TITLE
Pusher stops pushing, with lots of DNSLookupFailed errors

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#7113]
Our pusher process got behind. A couple of hours after it started to catch up, almost all pushes started failing:

![image](https://user-images.githubusercontent.com/1389908/77156873-31206000-6a98-11ea-8855-6102545355fa.png)

Inspection of logs showed that they were failing with DNSLookupFailed errors:

```
2020-03-19 11:11:27,083 - synapse.http.client - 281 - INFO - httppush.process-57923 - Sending request POST http://10.103.0.7/_matrix/push/v1/notify
2020-03-19 11:12:28,647 - synapse.http.client - 328 - INFO - httppush.process-57923 - Error sending request to  POST http://10.103.0.7/_matrix/push/v1/notify: DNSLookupError Couldn't find the hostname '10.103.0.7'
2020-03-19 11:12:28,647 - synapse.push.httppusher - 391 - WARNING - httppush.process-57923 - Failed to push event <redacted> to  <redacted>: <class 'twisted.internet.error.DNSLookupError'> DNS lookup failed: Couldn't find the hostname '10.103.0.7'.
2020-03-19 11:12:32,225 - synapse.push.httppusher - 276 - INFO - httppush.process-57923 - Push failed: delaying for 1s
```

**Comment by richvdh** [comment]
(it's worth pointing out that the matrix-org-hotfixes branch has this substitution to make it talk over http: https://github.com/matrix-org/synapse/blob/bc936b56572aef4fc16fffaa44a75bfdf99eeec4/synapse/push/httppusher.py#L106-L109)

**Comment by richvdh** [comment]
so here's what's going on

* each HTTP hit has a timeout of 60seconds
* we limit the number of concurrent DNS lookups to 100
* we end up doing so many DNS lookups that the queue grows so long that the DNS lookup does not complete within 60 seconds
* so the request gets cancelled
* but the DNS lookup does not
* pretty soon we end up spending all our time doing DNS lookups whose results will get thrown away

and yes, we're doing DNS lookups despite the fact it's an IP address

It appears that we're using Twisted's [GAIResolver](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/_resolver.py#L71), which works via `getaddrinfo(3)` calls from a threadpool. (It remains a mystery to me when we use the GAIResolver and when we use `twisted.names`).

**Comment by richvdh** [comment]
Some thoughts on what we could do about this:
* obviously, IP addresses don't need to have a DNS lookup. We could optimise this (it seems like it belongs in `HostnameEndpoint`), but it doesn't help the situation for anyone that's not `matrix.org`, and anyway I'd rather we weren't hardcoding the IP address of our HTTP gateway in the sourcecode.
* It would be nice if cancelling a request meant that the DNS lookup got cancelled. This would mean extending the `IHostnameResolver` interface in twisted.
* It would be nice if the DNS resolution didn't delegate via a threadpool. One way to do this might be to use twisted.names.client (cf #5053). Another might be to wrap `getaddrinfo_a` on platforms that support it.
* Possibly we should be limiting the number of pushes we attempt to do in parallel. One way to do this would be to keep track of the number of `HttpPusher` instances which are in their `_unsafe_process` loop, and once it reaches a certain level (100 or so?), stick them in a queue instead.
* Can we just define an update push API that lets you request more than one push per http hit?

**Comment by Kegsay** [comment]
> Our pusher process got behind. 

Isn't this embarrassingly parallelisable though? Each request does not depend on any other, so can't we just spin more processes up depending on load? Even if we did do the proposed fixes, how did it get so far behind in the first place? Slow push is also a huge problem.

**Comment by richvdh** [comment]
> Isn't this embarrassingly parallelisable though? Each request does not depend on any other, so can't we just spin more processes up depending on load? 

This is a bit simplistic. If you have multiple processes doing the pushing, you have to agree a way to make sure that each push gets processed by exactly one pusher. Obviously there are ways to solve that problem, but just spinning up more processes isn't likely to work currently.

>  Even if we did do the proposed fixes, how did it get so far behind in the first place? 

Since you ask, https://github.com/matrix-org/synapse/issues/7075 originally (we got two weeks behind). But actually this can happen whenever we have a brief period of a high volume of pushing to do, which the pusher can't keep up with. Once we start failing, the pusher drowns itself in retries and failures.

> Slow push is also a huge problem.

It is, but I'd rather have pushes happening 10 seconds late than a flurry of pushes making the entire process fall over so that pushes don't happen at all until someone manually restarts the thing.

**Comment by richvdh** [comment]
> Possibly we should be limiting the number of pushes we attempt to do in parallel. One way to do this would be to keep track of the number of HttpPusher instances which are in their _unsafe_process loop, and once it reaches a certain level (100 or so?), stick them in a queue instead.

this is definitely a thing. One of the things that can lead us into a death spiral is that, if the push gateway starts responding slowly for any reason, we can stack up thousands of concurrent connections, each of which take a long time to service. Indeed, they can take so long to service that the push gateway starts timing out the requests, which leads to us retrying them, which leads to more load, etc etc.

**Comment by o0laxkilla0o** [comment]
I believe this is effecting my homeserver as well. Is there anything I can do to rectify the situation?

**Comment by richvdh** [comment]
> (It remains a mystery to me when we use the GAIResolver and when we use `twisted.names`).

I'm going to write this down here because this seems to be the most likely place future me will find it.

* the `reactor.nameResolver` (and `reactor.resolver`) interfaces use `GAIResolver` (wrapped in our case by `_LimitedHostnameResolver` and in some cases `IPBlacklistingResolver`). This is installed in `ReactorBase.__init__`.
* however, we use `twisted.names` from `srv_resolver` (since the `IHostnameResolver` interface doesn't support SRV record lookups. (Incidentally, the equivalent of `getaddrinfo` seems to be `res_nquery`.)

**Comment by richvdh** [comment]
As I've just written in 44e88f80e, I don't really understand why we have a `_LimitedHostnameResolver`. (for links: that was added in #5037)

**Comment by valvanet** [comment]
I'm having problems resolving any domains in my docker setup since months now.. struggling to find any solution, neither federation, nor push or mail sending works anymore..

all my other containers work fine in terms of dns lookups. 

I have a docker network with nginx proxying, both 10.0.0.0 (host) and 172.0.0.0 (docker) networks involved.. but 
I also already tried with an empty blacklist and all whitelisted in my homeserver.yaml.. without success.

Push Services stopped working:
```
synapse.http.client - 452 - INFO - httppush.process-19 - Error sending request to  POST https://matrix.org/_matrix/push/v1/notify: DNSLookupError no results for hostname lookup: matrix.org
synapse.push.httppusher - 393 - WARNING - httppush.process-19 - Failed to push event $[...] to @root:domain.tld/im.vector.app.android/[...]: <class 'twisted.internet.error.DNSLookupError'> DNS lookup failed: no results for hostname lookup: matrix.org.
```

Mail sending stopped working:

```
synapse.handlers.identity - 415 - ERROR - POST-180 - Error sending threepid validation email to admin@domain.tld
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/identity.py", line 413, in send_threepid_validation
    await send_email_func(email_address, token, client_secret, session_id)
  File "/usr/local/lib/python3.8/site-packages/synapse/push/mailer.py", line 201, in send_add_threepid_mail
    await self.send_email(
  File "/usr/local/lib/python3.8/site-packages/synapse/push/mailer.py", line 318, in send_email
    await self.send_email_handler.send_email(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/send_email.py", line 175, in send_email
    await self._sendmail(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/send_email.py", line 111, in _sendmail
    await make_deferred_yieldable(d)
twisted.internet.error.DNSLookupError: **DNS lookup failed: mail.domain.tld**.
```

I finally came to the conclusion there must be kind of a bug in my setup or even in the used libraries.. 

Im not so firm with the "twisted" library, nor with python itself, but is there any workaround or clou how to find out if I'm affected by any bugs?

my matrix instance is dead since.
hope for help

o/

**Comment by xundeenergie** [comment]
I run my own push-gateway (Common-proxies from UnifiedPush), and notifications do not work, through this gateway.

Now i checked the log from matrix and get
`2022-01-28 11:26:54,862 - synapse.push.httppusher - 394 - WARNING - httppush.process-21 - Failed to push event $xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx to @jakob:schuerz.at/de.spiritcroc.riotx.up/https://ntfy.schuerz.at/xxxxxxxxxxxxxx?up=1: <class 'twisted.internet.error.DNSLookupError'> DNS lookup failed: no results for hostname lookup: ntfy.schuerz.at.`

But dig, host, nslookup can all resolve ntfy.schuerz.at correctly on internal dns and on external ones...

**Comment by reivilibre** [comment]
@xundeenergie Wonder if it's because it's a CNAME? However this issue is about push failing under load; if your push gateway never worked at all then it would be worth having a separate issue.

**Comment by StefanBalea** [comment]
In my case, the same error was received when the ip of the push server was added in ip_range_blacklist config. Removing the IP from this list fixed the problem.
